ogr/ogrsf_frmts/gpkg/ogrgeopackagedatasource.cpp

```diff
@@ -7324,6 +7324,273 @@ bool GDALGeoPackageDataset::DeleteVectorOrRasterLayer(const char *pszLayerName)
     return false;
 }
 
+bool GDALGeoPackageDataset::RenameVectorOrRasterLayer(
+    const char *pszLayerName, const char *pszNewLayerName)
+{
+    int idx = FindLayerIndex(pszLayerName);
+    if (idx >= 0)
+    {
+        m_papoLayers[idx]->Rename(pszNewLayerName);
+        return true;
+    }
+
+    char *pszSQL =
+        sqlite3_mprintf("SELECT 1 FROM gpkg_contents WHERE "
+                        "lower(table_name) = lower('%q') "
+                        "AND data_type IN ('tiles', '2d-gridded-coverage')",
+                        pszLayerName);
+    const bool bIsRasterTable = SQLGetInteger(hDB, pszSQL, nullptr) == 1;
+    sqlite3_free(pszSQL);
+
+    if (bIsRasterTable)
+    {
+        return RenameRasterLayer(pszLayerName, pszNewLayerName);
+    }
+
+    return false;
+}
+
+bool GDALGeoPackageDataset::RenameRasterLayer(const char *pszLayerName,
+                                              const char *pszNewLayerName)
+{
+    std::string osSQL;
+
+    char *pszSQL = sqlite3_mprintf(
+        "SELECT 1 FROM sqlite_master WHERE lower(name) = lower('%q') "
+        "AND type IN ('table', 'view')",
+        pszNewLayerName);
+    const bool bAlreadyExists = SQLGetInteger(GetDB(), pszSQL, nullptr) == 1;
+    sqlite3_free(pszSQL);
+    if (bAlreadyExists)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Table %s already exists",
+                 pszNewLayerName);
+        return false;
+    }
+
+    // Temporary remove foreign key checks
+    const GPKGTemporaryForeignKeyCheckDisabler
+        oGPKGTemporaryForeignKeyCheckDisabler(this);
+
+    if (SoftStartTransaction() != OGRERR_NONE)
+    {
+        return false;
+    }
+
+    osSQL = sqlite3_mprintf("UPDATE gpkg_contents SET table_name = '%q' WHERE "
```

Comment:
all below uses of sqlite3_mprintf() will leak. The return value needs to be freed with sqlite3_free

Reply:
done